sqlx-mysql/src/protocol/connect/auth_switch.rs

```diff
@@ -27,14 +27,23 @@ impl Decode<'_> for AuthSwitchRequest {
         let plugin = buf.get_str_nul()?.parse()?;
 
         // See: https://github.com/mysql/mysql-server/blob/ea7d2e2d16ac03afdd9cb72a972a95981107bf51/sql/auth/sha2_password.cc#L942
-        if buf.len() != 21 {
-            return Err(err_protocol!(
-                "expected 21 bytes but found {} bytes",
-                buf.len()
-            ));
-        }
-        let data = buf.get_bytes(20);
-        buf.advance(1); // NUL-terminator
+        let data = if buf.len() != 21 {
+            if matches!(plugin, AuthPlugin::MySqlClearPassword) {
+                // Contrary to the MySQL protocol, AWS Aurora with IAM sends
+                // no data. That is fine because the MySQL protocol says to
+                // ignore any data sent.
```

Comment:
Can we link documentation that corroborates these comments? Like https://dev.mysql.com/doc/dev/mysql-server/latest/page_protocol_connection_phase_authentication_methods_clear_text_password.html
or
https://dev.mysql.com/doc/refman/8.0/en/cleartext-pluggable-authentication.html

If the `MySqlClearPassword` protocol wants you to ignore data - and we're making a specific exception for the aurora no-data scenario, perhaps we can make that more clear like this.

```
if matches!(plugin, AuthPlugin::MySqlClearPassword) && buf.is_empty() {
  // Contrary to the MySQL protocol, AWS Aurora with IAM sends no data.
  // In this scenario, it is ok to proceed with an empty data buffer.
  return Ok(Self { plugin, data: Bytes::new() })
}
```

And then have the rest of the code left the way it is.